### PR TITLE
Cuda instance_norm fix

### PR DESCRIPTION
**Pull request by RyanUnderhill**
**Description**: Fix allocation size & initialize the values

### CODE REVIEW

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc [hunk 1]
```diff
@@ -100,10 +100,15 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
     CudnnTensor stats_desc;
     ORT_RETURN_IF_ERROR(stats_desc.Set(std::array<int64_t, 4>{1, stats_count, 1, 1}, CudnnTensor::GetDataType<CudaT>()));
 
-    auto mean = GetScratchBuffer<CudaT>(stats_count);
-    auto variance = GetScratchBuffer<CudaT>(stats_count);
-    auto unused_scale = GetScratchBuffer<CudaT>(stats_count);
-    auto unused_bias = GetScratchBuffer<CudaT>(stats_count);
+    auto mean = GetScratchBuffer<CudaT>(stats_count * sizeof(T));
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:103]: Can ```stats_count * sizeof(T)``` be done only once?

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc [hunk 2]
```diff
@@ -100,10 +100,15 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
...
+    auto variance = GetScratchBuffer<CudaT>(stats_count * sizeof(T));
+    auto unused_scale = GetScratchBuffer<CudaT>(stats_count * sizeof(T));
+    auto unused_bias = GetScratchBuffer<CudaT>(stats_count * sizeof(T));
+
+    cudaMemset(mean.get(), 0, stats_count * sizeof(T));
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:108]: `cudaMemsetAsync()` with the session's stream `(Stream())` ?

**Comment by hariharans29** [comment]
Possibly fixes this as well : https://github.com/microsoft/onnxruntime/issues/8766

**Review by hariharans29 — APPROVED** [review]
LGTM

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc [hunk 3]
```diff
@@ -100,10 +100,19 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
     CudnnTensor stats_desc;
     ORT_RETURN_IF_ERROR(stats_desc.Set(std::array<int64_t, 4>{1, stats_count, 1, 1}, CudnnTensor::GetDataType<CudaT>()));
 
-    auto mean = GetScratchBuffer<CudaT>(stats_count);
-    auto variance = GetScratchBuffer<CudaT>(stats_count);
-    auto unused_scale = GetScratchBuffer<CudaT>(stats_count);
-    auto unused_bias = GetScratchBuffer<CudaT>(stats_count);
+    size_t stats_buffer_size = stats_count * sizeof(T);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:103]: >sizeof(T) [](http://example.com/codeflow?start=45&length=9)

sizeof(CudaT) ?

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc [hunk 4]
```diff
@@ -100,10 +100,19 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
...
+
+    auto mean = GetScratchBuffer<CudaT>(stats_buffer_size);
+    cudaMemsetAsync(mean.get(), 0, stats_buffer_size, Stream());
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:106]: >udaMemsetAsync(mean.get(), 0, stats_buffer_size, Stream() [](http://example.com/codeflow?start=5&length=57)

These are potentially expensive. Please, document the justification.

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 2 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: Indeed, fixed

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 1 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: Yep, fixed

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 3 above)
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:103]: >siz [](http://example.com/codeflow?start=4&length=3)

const

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc [hunk 5]
```diff
@@ -100,10 +100,19 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
     CudnnTensor stats_desc;
     ORT_RETURN_IF_ERROR(stats_desc.Set(std::array<int64_t, 4>{1, stats_count, 1, 1}, CudnnTensor::GetDataType<CudaT>()));
 
-    auto mean = GetScratchBuffer<CudaT>(stats_count);
-    auto variance = GetScratchBuffer<CudaT>(stats_count);
-    auto unused_scale = GetScratchBuffer<CudaT>(stats_count);
-    auto unused_bias = GetScratchBuffer<CudaT>(stats_count);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:106]: >GetScratchBuffer [](http://example.com/codeflow?start=23&length=16)

This interface is terribly wrong. It's argument names count_or_bytes. It is templated and it returns a typed smartptr, all of which is two big lies. And an invitation to pass the count of items.
It requires size in bytes and returns a bytes allocated buffer. It should never been a template.

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 1 above)
Comment (yuslepukhin) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: I just looked into the implementation and the usages of this function. It actually expects a count of typed items. Need more investigation.

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 4 above)
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:106]: >cudaMemsetAsync [](http://example.com/codeflow?start=4&length=15)

Also this needs CUDA_RETURN_IF_ERROR

Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: Documented

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 3 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: Good catch

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 4 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: And now it has it

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc (hunk 3 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/nn/instance_norm.cc]: Consted

**Review by hariharans29 — APPROVED** [review]

File: onnxruntime/core/providers/cuda/nn/instance_norm.cc [hunk 6]
```diff
@@ -100,10 +100,19 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
     CudnnTensor stats_desc;
     ORT_RETURN_IF_ERROR(stats_desc.Set(std::array<int64_t, 4>{1, stats_count, 1, 1}, CudnnTensor::GetDataType<CudaT>()));
 
+    const size_t stats_byte_count = stats_count * sizeof(CudaT);
+
+    // Mean & Variance are inputs & outputs and must be initialized to zero to work properly
     auto mean = GetScratchBuffer<CudaT>(stats_count);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/nn/instance_norm.cc:106]: >GetScratchBuffer [](http://example.com/codeflow?start=16&length=16)

I would suggest to allocate one big buffer and set to to zero once. Then pass ptrs to different parts of the buffer there.

**Pull request merged**